lib/events/dynamoevents/dynamoevents.go

```diff
@@ -499,46 +500,58 @@ func (l *Log) SearchEvents(fromUTC, toUTC time.Time, filter string, limit int) (
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	input := dynamodb.QueryInput{
-		KeyConditionExpression:    aws.String(query),
-		TableName:                 aws.String(l.Tablename),
-		ExpressionAttributeValues: attributeValues,
-		IndexName:                 aws.String(indexTimeSearch),
-	}
-	start := time.Now()
-	out, err := l.svc.Query(&input)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	g.WithFields(log.Fields{"duration": time.Since(start), "items": len(out.Items)}).Debugf("Query completed.")
+
+	var lastEvaluatedKey map[string]*dynamodb.AttributeValue
 	var total int
-	for _, item := range out.Items {
-		var e event
-		if err := dynamodbattribute.UnmarshalMap(item, &e); err != nil {
-			return nil, trace.BadParameter("failed to unmarshal event for %v", err)
+	for i := 0; i < backend.DefaultLargeLimit/100; i++ {
+		input := dynamodb.QueryInput{
+			KeyConditionExpression:    aws.String(query),
+			TableName:                 aws.String(l.Tablename),
+			ExpressionAttributeValues: attributeValues,
+			IndexName:                 aws.String(indexTimeSearch),
+			ExclusiveStartKey:         lastEvaluatedKey,
 		}
-		var fields events.EventFields
-		data := []byte(e.Fields)
-		if err := json.Unmarshal(data, &fields); err != nil {
-			return nil, trace.BadParameter("failed to unmarshal event %v", err)
+		start := time.Now()
+		out, err := l.svc.Query(&input)
+		if err != nil {
+			return nil, trace.Wrap(err)
 		}
-		var accepted bool
-		for i := range eventFilter {
-			if fields.GetString(events.EventType) == eventFilter[i] {
-				accepted = true
-				break
+		g.WithFields(log.Fields{"duration": time.Since(start), "items": len(out.Items)}).Debugf("Query completed.")
+
+		for _, item := range out.Items {
+			var e event
+			if err := dynamodbattribute.UnmarshalMap(item, &e); err != nil {
+				return nil, trace.BadParameter("failed to unmarshal event for %v", err)
 			}
-		}
-		if accepted || !doFilter {
-			values = append(values, fields)
-			total++
-			if limit > 0 && total >= limit {
-				break
+			var fields events.EventFields
+			data := []byte(e.Fields)
+			if err := json.Unmarshal(data, &fields); err != nil {
+				return nil, trace.BadParameter("failed to unmarshal event %v", err)
+			}
+			var accepted bool
+			for i := range eventFilter {
+				if fields.GetString(events.EventType) == eventFilter[i] {
+					accepted = true
+					break
+				}
 			}
+			if accepted || !doFilter {
+				values = append(values, fields)
+				total++
+				if limit > 0 && total >= limit {
+					break
+				}
+			}
+		}
+
+		lastEvaluatedKey = out.LastEvaluatedKey
+		if len(lastEvaluatedKey) == 0 {
```

Comment:
A comment explaining that empty LastEvaluatedKey means the last page of results was returned wouldn't hurt (I had to go to the docs to confirm).